### ISSUE TITLE
optimizer: pk lost while optimize filter

### ISSUE THREAD

**Issue by skyzh** [singularity-data/risingwave#3118]
<img width="790" alt="image" src="https://user-images.githubusercontent.com/4198311/172989896-fee86496-af2d-4702-809f-f2c265fe495b.png">

The StreamProject doesn't contain pk v3

**Comment by xiangjinwu** [comment]
Repro: cc @neverchanje 
```
create table t (v1 varchar, v2 varchar, v3 varchar);

create materialized view mv as
select
    *
from
    t
group by
    v1,
    v2,
    v3;

create materialized view mv2 as
select
    max(v1) as v1
from
    mv
where
    v3 = 'world' or v3 = 'hello';

insert into t values ('111', '222', 'hello'), ('112', '222', 'hello');

insert into t values ('111', '222', 'world'), ('112', '222', 'world');  -- if 4 tuples are inserted together, the same panic "invalid flush status: double insert"

select * from mv2;

delete from t where v1 = '112' and v2 = '222' and v3 = 'world';

select * from mv2;  -- expected max is 112 but actual max is 111
```

**Comment by lmatz** [comment]
Seems there are some other problems:

![SCR-20220612-vp7](https://user-images.githubusercontent.com/5791930/173238926-bca74cf3-f87e-4f5c-a9d4-1f8117af3462.png)

Is it possible that we first do `predicate_pull_up` to ensure that after the PK derivation, there will be no more or fewer operators in the plan? 
Otherwise, it seems that we need to handle extra cases here and there when more operators are created. 
But I am not that familiar with the mechanism of each pass in the optimization and their interrelations now, it could be no difference.

**Comment by skyzh** [comment]
I think we can simply retain pk when doing column prune for filter, not sure whether this would work...

**Comment by skyzh** [comment]
Oops, I see. It seems that pk is added in stream phase. In logical phase we will prune regardless of pk.

Basically the pk is added here:

https://github.com/singularity-data/risingwave/blob/3acee84536e495659e024fc3730681128bc8212b/src/frontend/src/optimizer/plan_node/logical_scan.rs#L373-L375

But not all cases need stream pk. If we only have stateless operators, we can throw away some pks.

The first thing come into my mind is to fix the add pk process, and add all pks instead of only the required ones. Then we can do a column prune to remove unnecessary ones. But I'm not sure whether this would work.

Anyway, to do fewer hacks, I think we can even scan pk even for batch queries, so that we will always need to add pk in logical phase.

**Comment by st1page** [comment]
sorry but why we need #3119 ? I think the LogicalProject::rewrite_for_stream should add the pk then in the stream phase we should maintain the pk for any operation.

**Comment by lmatz** [comment]
> If we only have stateless operators, we can throw away some pks.

As the MV operator is included in a streaming plan, I believe it always needs to have all the pk scanned?

**Comment by skyzh** [comment]
> sorry but why we need #3119 ? I think the LogicalProject::rewrite_for_stream should add the pk then in the stream phase we should maintain the pk for any operation.

Yes, that's a quick fix specifically for #3086. If we can include all pks in to_stream, then we can remove such workarounds.
